kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -141,31 +154,56 @@ private ResultOrError handleNext(T result, @Nullable Throwable error) {
   private static class InputStreamingResponse<T> extends StreamingResponse<T> {
 
     private final MappingIterator<T> mappingIteratorInput;
+    private final StreamingResponseIdleTimeLimiter idleTimeLimiter;
+
+    private Throwable hasNextError = null;
+    private boolean closed = false;
 
     private InputStreamingResponse(
-        MappingIterator<T> mappingIteratorInput, ChunkedOutputFactory chunkedOutputFactory) {
+        MappingIterator<T> mappingIteratorInput,
+        ChunkedOutputFactory chunkedOutputFactory,
+        StreamingResponseIdleTimeLimiter idleTimeLimiter) {
       super(chunkedOutputFactory);
       this.mappingIteratorInput = requireNonNull(mappingIteratorInput);
+      this.idleTimeLimiter = requireNonNull(idleTimeLimiter);
     }
 
     public void close() {
-      try {
-        mappingIteratorInput.close();
-      } catch (IOException e) {
-        log.error("Error when closing the request stream.", e);
-      }
+      closed = true;
     }
 
     @Override
     public boolean hasNext() {
-      return mappingIteratorInput.hasNext();
+      if (closed) {
+        return false;
+      }
+      try {
+        return idleTimeLimiter.callWithTimeout(mappingIteratorInput::hasNextValue);
```

Comment:
I'd like to check my understanding of the subtle change regarding the checked vs the unchecked version of the hasNext idiom here. If something checked is thrown here, it would still be transformed into an `ExecutionException`, just like anything unchecked, right? Also the unchecked version uses the checked one behind the scenes, so using the checked one is slightly more efficient. Otherwise the two are interchangeable here. Is that correct?

Reply:
I think unchecked throws `UncheckedExecutionException`, and by "I think" I mean "i've seen it when doing manual tests". :)

Reply:
Well, it's not the unchecked version of the hasNext idiom, is the subsequent wrapping in Guava, right?
I should have written just _execution exception_ because one is not a subtype of the other - what I meant is no matter what we use, we'll have any exceptions wrapped in the corresponding form of an _execution exception_, and no matter what form of an _execution exception_ is thrown, this change would have treated it the same (because both the checked and the unchecked exceptions are always handled, and always handled in the same `catch`).
So am I right to assume there is no reason to choose one over the other besides avoiding the redundant wrapping on both the Jackson side and the Guava side?